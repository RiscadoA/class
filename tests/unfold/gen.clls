type corec Gen(T) {
    offer of {
        |#stop: close
        |#next: send T; Gen(T)
    }
};;

proc rec close_g(g: Gen(close)) {
    println("unfold g"); unfold g;
    println("case"); case g of {
        |#stop: println("close g"); close g
        |#next:
            println("send"); send g(x. println("close"); close x);
            close_g(g)
    }
};;

proc main() {
    cut {
        println("unfold 1"); unfold g;
        println("#next 1"); #next g;
        println("recv 1"); recv g(x);
        println("wait 1"); wait x;
        println("unfold 2"); unfold g;
        println("#next 2"); #next g;
        println("recv 2"); recv g(y);
        println("wait 2"); wait y;
        println("unfold 3"); unfold g;
        println("stop"); #stop g;
        println("wait g"); wait g;
        ()
        |g: Gen(close)|
        close_g(g)
    }
};;
