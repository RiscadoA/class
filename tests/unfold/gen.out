unfold 1
unfold g
#next 1
case
send
close
recv 1
wait 1
unfold 2
unfold g
#next 2
case
send
close
recv 2
wait 2
unfold 3
unfold g
stop
case
close g
wait g
